# ai_project_helper/client/operations/execute_plan.py

import grpc
import os
from datetime import datetime
from ai_project_helper.proto import helper_pb2, helper_pb2_grpc
from ai_project_helper.client.utils import (
    save_execution_log,
    print_feedback,
    init_statistics,
    truncate_long_text,
    print_summary
)

def run_get_plan_then_execute(request, context):
    """æ‰§è¡Œè·å–è®¡åˆ’å¹¶ç«‹å³æ‰§è¡Œæ“ä½œ"""
    logger = context["logger"]
    statistics = init_statistics()
    start_time = datetime.now()
    complete_plan = ""
    execution_log = ""
    
    with grpc.insecure_channel(context["grpc_channel"]) as channel:
        stub = helper_pb2_grpc.AIProjectHelperStub(channel)
        
        logger.info(f"ğŸ“ è¯·æ±‚ç”Ÿæˆå¹¶æ‰§è¡Œè®¡åˆ’: {request.requirement}")
        
        try:
            for feedback in stub.GetPlanThenRun(request):
                # æ”¶é›†å®Œæ•´æ—¥å¿—
                log_entry = f"Step [{feedback.step_index}/{feedback.total_steps}] - {feedback.step_description}\n"
                if feedback.output:
                    log_entry += f"è¾“å‡º: {feedback.output}\n"
                if feedback.error:
                    log_entry += f"é”™è¯¯: {feedback.error}\n"
                execution_log += log_entry + "-" * 60 + "\n"
                
                print_feedback(feedback)
                
                # ç»Ÿè®¡è®¡åˆ’éƒ¨åˆ†
                if feedback.action_index < 0:
                    statistics["plan_parts"] += 1
                
                # ä¿å­˜å®Œæ•´è®¡åˆ’
                if feedback.complete_plan:
                    complete_plan = feedback.complete_plan
                    save_plan(request.project_id, complete_plan)
                
                # åªç»Ÿè®¡æ‰§è¡ŒåŠ¨ä½œçš„æœ€ç»ˆçŠ¶æ€
                if feedback.action_index >= 0 and feedback.status.lower() in ["success", "warning", "failed"]:
                    statistics["total_actions"] += 1
                    statistics["action_types"][feedback.action_type] += 1
                    
                    # æ›´æ–°æ­¥éª¤è®¡æ•°
                    if feedback.step_index > statistics["total_steps"]:
                        statistics["total_steps"] = feedback.step_index
                    
                    # è®°å½•é—®é¢˜ä¿¡æ¯
                    if feedback.status.lower() == "warning":
                        statistics["warning_actions"] += 1
                        statistics["warnings"].append({
                            "step": feedback.step_index,
                            "action": feedback.action_index + 1,
                            "description": feedback.step_description,
                            "message": feedback.error or feedback.output
                        })
                    elif feedback.status.lower() == "failed":
                        statistics["failed_actions"] += 1
                        statistics["errors"].append({
                            "step": feedback.step_index,
                            "action": feedback.action_index + 1,
                            "description": feedback.step_description,
                            "message": feedback.error
                        })
                    else:  # success
                        statistics["success_actions"] += 1
                        
        except grpc.RpcError as e:
            logger.error(f"gRPCé”™è¯¯: {e.code()}: {e.details()}")
            statistics["errors"].append({
                "step": "é€šä¿¡é”™è¯¯",
                "action": "N/A",
                "description": "gRPCé€šä¿¡å¤±è´¥",
                "message": f"{e.code()}: {e.details()}"
            })
            statistics["failed_actions"] += 1
    
    # ä¿å­˜æ‰§è¡Œæ—¥å¿—
    save_execution_log(request.project_id, execution_log)
    
    duration = (datetime.now() - start_time).total_seconds()
    print_summary(statistics, duration)
    return statistics